server/src-lib/Hasura/RQL/DDL/Schema/Rename.hs

```diff
@@ -269,11 +287,38 @@ updateCols qt rf permSpec =
       PCCols c -> PCCols $ flip map c $
         \col -> if col == oCol then nCol else col
 
-updateBoolExp
+updateTableInBoolExp :: RenameTable -> BoolExp -> BoolExp
+updateTableInBoolExp rt@(oldQT, newQT) (BoolExp bExp) = BoolExp $
+  case bExp of
```

Comment:
Can't able to use `traverse` since we've to deal `BEExists` value differently. This is followed everywhere in this PR. Suggest any other better way of implementing this.

Reply:
It took me a little bit of researching to figure out how to do this, but yes, `lens` can help here.

There are a handful of different ways to do this, but the easiest way is to use [`Control.Lens.Plated`](https://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Plated.html). That module provides a variety of functions for operating on *self-similar* data structures. [`Data.Data.Lens`](https://hackage.haskell.org/package/lens-4.17.1/docs/Data-Data-Lens.html) includes related functionality, via the [`template`](https://hackage.haskell.org/package/lens-4.17.1/docs/Data-Data-Lens.html#v:template) traversal, which finds all occurrences of a particular type in a data structure. However, `template` does not handle the *self-similar* case, where the type you’re looking for is inside itself—it only stops on the outermost occurrence of the type you’re looking for.

In contrast, `Control.Lens.Plated` provides an extremely powerful operation, [`transform`](https://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Plated.html#v:transform), which applies a function to every element of a self-similar tree, bottom-up. We can use this operation to implement these functions: we just find all the `GBoolExp`s inside each `BoolExp`, recursively, then apply an update function to each of them.

To do this, we’ll need to define `Plated` instances for `GBoolExp` and `GExists`, since those are the types we care about recursively traversing. However, if a `Data` instance exists already, then the body of a `Plated` instance can be empty, as it will default to using the [`uniplate`](https://hackage.haskell.org/package/lens-4.17.1/docs/Data-Data-Lens.html#v:uniplate) operation from `Data.Data.Lens`. Therefore, writing these instances is trivial:

```haskell
instance (Data a) => Plated (GExists a)
instance (Data a) => Plated (GBoolExp a)
```

We’ll also want to derive some lenses, prisms, and isomorphisms:

```haskell
$(makeLenses ''GExists)
$(makePrisms ''GBoolExp)
$(makeWrapped ''BoolExp)
```

Now we can write `updateTableInBoolExp` in just a couple lines of code:

```haskell
updateTableInBoolExp :: RenameTable -> BoolExp -> BoolExp
updateTableInBoolExp (oldQT, newQT) =
  over _Wrapped . transform $ _BoolExists.geTable %~ \rqfQT ->
    if rqfQT == oldQT then newQT else rqfQT
```

Here’s how it works:

  1. We start by using `over _Wrapped` to move “under” the `BoolExp` wrapper. (We derived a `Wrapped` instance for `BoolExp` using `makeWrapped`, which is what allows us to use `_Wrapped`.)

  2. We call `transform` to apply a function to all the self-similar components of `GBoolExp ColExp`.

  3. We use `_BoolExists.geTable` to get a `Traversal` that targets each of the tables inside the `GBoolExp ColExp`.

  4. We use `%~` (which is actually just an infix alias for `over`) to apply a function to each of the tables targeted by the traversal.

That’s it! `transform` is doing the heavy lifting here, working out how to do all the recursion for us.

---

Unfortunately, we can’t use `Control.Lens.Plated` to implement `updateFieldInBoolExp`, since the values we’re updating aren’t actually self-similar at the type level. However, that’s not actually a problem, since we don’t actually do anything special in the `BoolExists` case. Therefore, we can just keep using the implementation that uses `traverse` and explicit mutual recursion, and everything should continue to work:

```haskell
updateFieldInBoolExp
  :: (QErrM m, CacheRM m)
  => QualifiedTable -> RenameField -> BoolExp -> m BoolExp
updateFieldInBoolExp qt rf = fmap BoolExp . traverse (updateColExp qt rf) . unBoolExp
```

Reply:
In `updateFieldInBoolExp` function:
In the `BoolExists` case, the reference table changes. So, we need to specially handle that case.

Reply:
Refer to `updateColExp` in `BoolFld` case. It takes the reference table and updates the columns.